### PR TITLE
Add companion notebook for Forecast + SageMaker NTM Blog

### PR DESCRIPTION
**Pull request by dehrlich**
Uploaded folder containing jupyter notebooks, util, and README files that accompany the AWS blog "Time series forecasting using unstructured data with Amazon Forecast and Amazon SageMaker Neural Topic Model" that is currently in tech review. This collection of notebooks has already undergone tech review by the Forecast team and the edits have been incorporated.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by shimomut** [comment]
I am starting reviewing this pull request.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/1_preprocess.ipynb [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/00432/Data/News_Final.csv'\n",
+    "r = requests.get(url, allow_redirects=True, verify=False)\n",
+    "open('data/News_Final.csv', 'wb').write(r.content)"
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/1_preprocess.ipynb:65]: The file is not closed after writing. Please close the file or use "with" statement to make sure file is closed.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb [hunk 2]
```diff
@@ -0,0 +1,838 @@
...
+    "import sys\n",
+    "import time\n",
+    "import re\n",
+    "import boto3\n",
+    "import sagemaker\n",
+    "import nltk\n",
+    "# importing forecast notebook utility from notebooks/ directory\n",
+    "sys.path.insert( 0, os.path.abspath(\"./util\") )\n",
+    "import util\n",
+    "\n",
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb:55]: Delete Delete "amazon-forecast-samples/notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/util", and use "amazon-forecast-samples/notebooks/common/util" instead.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb [hunk 3]
```diff
@@ -0,0 +1,972 @@
...
+    "from sklearn.metrics import mean_squared_error\n",
+    "import time\n",
+    "import datetime as dt\n",
+    "from time import sleep\n",
+    "import boto3\n",
+    "import sagemaker\n",
+    "# importing forecast notebook utility from notebooks/ directory\n",
+    "sys.path.insert( 0, os.path.abspath(\"./util\") )\n",
+    "import util"
+   ]
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb:50]: Delete Delete "amazon-forecast-samples/notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/util", and use "amazon-forecast-samples/notebooks/common/util" instead.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb [hunk 4]
```diff
@@ -0,0 +1,838 @@
...
+    "Another way to avoid overfitting is to introduce EarlyStopping which is done here by num_patience_epochs which ensure that the training is stopped if the change in the loss is less than the specified tolerance for a number of epochs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%time\n",
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb:475]: If this magic command is intended to measure the time required to execute this cell, should use "%%time" instead.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb [hunk 5]
```diff
@@ -0,0 +1,838 @@
...
+    "We also visualize the Headlines in a T-SNE plot to capture the number of distinct Topic clusters that appear."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%time\n",
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb:591]: If this magic command is intended to measure the time required to execute this cell, should use "%%time" instead.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb [hunk 6]
```diff
@@ -0,0 +1,972 @@
...
+    "    response = forecast.create_dataset(\n",
+    "                    Domain=\"WEB_TRAFFIC\",\n",
+    "                    DatasetType='TARGET_TIME_SERIES',\n",
+    "                    DatasetName=datasetName,\n",
+    "                    DataFrequency=DATASET_FREQUENCY, \n",
+    "                    Schema = schema\n",
+    "                   )\n",
+    "    datasetArn = response['DatasetArn']\n",
+    "    print('Success')\n",
+    "except Exception as e:\n",
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb:385]: Catching generic "Exception" class to skip creating existing resources, but shouldn't do this as it is possible to catch different types of errors.
Use forecast.exceptions.ResourceAlreadyExistsException specifically.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb [hunk 7]
```diff
@@ -0,0 +1,972 @@
...
+    "    response=forecast.create_dataset(\n",
+    "                    Domain=\"WEB_TRAFFIC\",\n",
+    "                    DatasetType='RELATED_TIME_SERIES',\n",
+    "                    DatasetName=datasetName,\n",
+    "                    DataFrequency=DATASET_FREQUENCY, \n",
+    "                    Schema = schema\n",
+    "                   )\n",
+    "    related_datasetArn = response['DatasetArn']\n",
+    "    print('Success')\n",
+    "except Exception as e:\n",
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb:437]: Catching generic "Exception" class to skip creating existing resources, but shouldn't do this as it is possible to catch different types of errors.
Use forecast.exceptions.ResourceAlreadyExistsException specifically.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb [hunk 8]
```diff
@@ -0,0 +1,972 @@
...
+    "    \n",
+    "try:\n",
+    "    create_dataset_group_response = forecast.create_dataset_group(DatasetGroupName=datasetGroupName,\n",
+    "                                                              Domain=\"WEB_TRAFFIC\",\n",
+    "                                                              DatasetArns= [datasetArn, related_datasetArn]\n",
+    "                                                             )\n",
+    "    datasetGroupArn = create_dataset_group_response['DatasetGroupArn']\n",
+    "    print('Success')\n",
+    "\n",
+    "except Exception as e:\n",
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb:472]: Catching generic "Exception" class to skip creating existing resources, but shouldn't do this as it is possible to catch different types of errors.
Use forecast.exceptions.ResourceAlreadyExistsException specifically.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb [hunk 9]
```diff
@@ -0,0 +1,972 @@
...
+    "                                                              \"S3Config\" : {\n",
+    "                                                                 \"Path\":s3DataPath,\n",
+    "                                                                 \"RoleArn\": role_arn\n",
+    "                                                              } \n",
+    "                                                          },\n",
+    "                                                          TimestampFormat=TIMESTAMP_FORMAT\n",
+    "                                                         )\n",
+    "    ds_import_job_arn=ds_import_job_response['DatasetImportJobArn']\n",
+    "    target_ds_import_job_arn = copy.copy(ds_import_job_arn) #used to delete the resource during cleanup\n",
+    "except Exception as e:\n",
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb:525]: Catching generic "Exception" class to skip creating existing resources, but shouldn't do this as it is possible to catch different types of errors.
Use forecast.exceptions.ResourceAlreadyExistsException specifically.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb [hunk 10]
```diff
@@ -0,0 +1,972 @@
...
+    "                                                              \"S3Config\" : {\n",
+    "                                                                 \"Path\":s3DataPath,\n",
+    "                                                                 \"RoleArn\": role_arn\n",
+    "                                                              } \n",
+    "                                                          },\n",
+    "                                                          TimestampFormat=TIMESTAMP_FORMAT\n",
+    "                                                         )\n",
+    "    ds_import_job_arn=ds_import_job_response['DatasetImportJobArn']\n",
+    "    related_ds_import_job_arn = copy.copy(ds_import_job_arn) #used to delete the resource during cleanup\n",
+    "except Exception as e:\n",
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb:567]: Catching generic "Exception" class to skip creating existing resources, but shouldn't do this as it is possible to catch different types of errors.
Use forecast.exceptions.ResourceAlreadyExistsException specifically.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb [hunk 11]
```diff
@@ -0,0 +1,972 @@
...
+    "                                                  PerformAutoML=False, # change to true if want to perform AutoML\n",
+    "                                                  PerformHPO=False, # change to true to perform HPO\n",
+    "                                                  EvaluationParameters= {\"NumberOfBacktestWindows\": 1, \n",
+    "                                                                         \"BackTestWindowOffset\": 15}, \n",
+    "                                                  InputDataConfig= {\"DatasetGroupArn\": datasetGroupArn},\n",
+    "                                                  FeaturizationConfig= {\"ForecastFrequency\": \"D\", \n",
+    "                                                                        }\n",
+    "                                                 )\n",
+    "    predictorArn=create_predictor_response['PredictorArn']\n",
+    "except Exception as e:\n",
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb:652]: Catching generic "Exception" class to skip creating existing resources, but shouldn't do this as it is possible to catch different types of errors.
Use forecast.exceptions.ResourceAlreadyExistsException specifically.

File: notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb [hunk 12]
```diff
@@ -0,0 +1,838 @@
...
+    "# This could be easily done by providing AmazonS3FullAccess but this is **not** a recommended security best practice.\n",
+    "# then input that role in the text widget below.\n",
+    "\n",
+    "try:\n",
+    "    from sagemaker import get_execution_role\n",
+    "    role = get_execution_role()\n",
+    "except Exception as e:\n",
+    "    role_widget = util.create_text_widget( \"IAM Role Name\", \"Enter the IAM role needed for SageMaker to access training containers\" )\n",
+    "    role = role_widget.value\n",
+    "    assert role, \"role not set.\""
```
Comment (shimomut) [inline notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb:390]: In local Jupyter environment, customers always get "AssertionError: role not set." when this cell is executed first time. 

Like other places, split the latter 2 lines as separate cell.

Alternatively, you can use Python's input() function. ( I recently found that input() is easier to accept customer's text input interactively than ipywidget's text widget.

**Review by shimomut — CHANGES_REQUESTED** [review]
Added my feedbacks.

File: notebooks/blog_materials/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/1_preprocess.ipynb [hunk 13]
```diff
@@ -62,7 +62,8 @@
    "source": [
     "url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/00432/Data/News_Final.csv'\n",
     "r = requests.get(url, allow_redirects=True, verify=False)\n",
-    "open('data/News_Final.csv', 'wb').write(r.content)"
+    "open('data/News_Final.csv', 'wb').write(r.content)\n",
+    "r.close()"
```
Comment (shimomut) [inline notebooks/blog_materials/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/1_preprocess.ipynb:66]: I meant the opened local file ("News_Final.csv") has to be closed after written, rather than closing the requests.response object.
r.close() may not be needed.

sample:

with open('data/News_Final.csv', 'wb') as fd:
    fd.write(r.content)

File: notebooks/advanced/Compare_Multiple_Models/Compare_Multiple_Models.ipynb [hunk 14]
```diff
@@ -517,7 +517,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.6.5"
+   "version": "3.7.3"
   }
```
Comment (shimomut) [inline notebooks/advanced/Compare_Multiple_Models/Compare_Multiple_Models.ipynb:521]: I suppose this change to unrelated notebook file was not intended.

File: notebooks/blog_materials/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/1_preprocess.ipynb [hunk 15]
```diff
@@ -265,7 +266,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.6.5"
+   "version": "3.7.3"
```
Comment (shimomut) [inline notebooks/blog_materials/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/1_preprocess.ipynb:269]: This is not a critical problem, but could you use old kernel spec ("conda_python3") and Python version, so that
1. customers don't see "Kernel not found" dialog when this notebook is opened on the SageMaker Jupyter environment.
2. consistency with other notebooks.

File: notebooks/blog_materials/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb [hunk 16]
```diff
@@ -830,7 +838,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.6.5"
+   "version": "3.7.3"
```
Comment (shimomut) [inline notebooks/blog_materials/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb:841]: This is not a critical problem, but could you use old kernel spec ("conda_python3") and Python version, so that
1. customers don't see "Kernel not found" dialog when this notebook is opened on the SageMaker Jupyter environment.
2. consistency with other notebooks.

File: notebooks/blog_materials/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb [hunk 17]
```diff
@@ -964,7 +964,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.6.5"
+   "version": "3.7.3"
```
Comment (shimomut) [inline notebooks/blog_materials/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/3_Forecast.ipynb:967]: This is not a critical problem, but could you use old kernel spec ("conda_python3") and Python version, so that
1. customers don't see "Kernel not found" dialog when this notebook is opened on the SageMaker Jupyter environment.
2. consistency with other notebooks.

**Review by shimomut — CHANGES_REQUESTED** [review]
Found 1 remaining unfixed problem ( file closure ), and some uninteded changes of notebook metadata.

File: notebooks/advanced/Compare_Multiple_Models/Compare_Multiple_Models.ipynb (hunk 14 above)
Comment (dehrlich) [reply notebooks/advanced/Compare_Multiple_Models/Compare_Multiple_Models.ipynb:521]: Yes you are right. I did not mean to change anything in that notebook. I opened it to look at how something had been implemented to get an idea of how to do it on the notebook I am currently working on. It must have updated the python version when I did a pip install --upgrade on my local branch, although I am not exactly sure how.

**Comment by shimomut** [comment]
Thank you for addressing all feedbacks. Merged.
